src/test/java/io/aiven/connect/elasticsearch/ElasticsearchSinkFailureTest.java

```diff
@@ -0,0 +1,109 @@
+            }
+            return BulkResponse.success();
+        });
+
+        final Map<String, String> props  = new HashMap<>();
+        props.put(ElasticsearchSinkConnectorConfig.TYPE_NAME_CONFIG, "kafka-connect");
+        props.put(ElasticsearchSinkConnectorConfig.CONNECTION_URL_CONFIG, "localhost");
+        props.put(ElasticsearchSinkConnectorConfig.KEY_IGNORE_CONFIG, "true");
+        props.put(ElasticsearchSinkConnectorConfig.DROP_INVALID_MESSAGE_CONFIG, "true");
+        props.put(ElasticsearchSinkConnectorConfig.MAX_RETRIES_CONFIG, "4");
+
+        elasticsearchSinkTask.start(props, failingClient);
+        elasticsearchSinkTask.put(Collections.singletonList(new SinkRecord("topic", 0, null, null, null, "foo", 0)));
+        elasticsearchSinkTask.flush(Collections.emptyMap());
+        elasticsearchSinkTask.put(Collections.singletonList(new SinkRecord("topic", 0, null, null, null, "bar", 0)));
+        elasticsearchSinkTask.flush(Collections.emptyMap());
+        // 3 retries + 1 success
+        assertEquals(4, apiCallCounter.get());
+    }
+
+    @Test
+    public void testRaiseExceptionIfNot() throws IOException {
+        final ElasticsearchSinkTask elasticsearchSinkTask = new ElasticsearchSinkTask();
+        final JestElasticsearchClient failingClient = Mockito.mock(JestElasticsearchClient.class);
+        final AtomicInteger apiCallCounter = new AtomicInteger(0);
+        when(failingClient.executeBulk(any())).thenAnswer(i -> {
+            apiCallCounter.incrementAndGet();
+            return BulkResponse.failure(false, "");
+        });
+
+        final Map<String, String> props = new HashMap<>();
+        props.put(ElasticsearchSinkConnectorConfig.TYPE_NAME_CONFIG, "kafka-connect");
+        props.put(ElasticsearchSinkConnectorConfig.CONNECTION_URL_CONFIG, "localhost");
+        props.put(ElasticsearchSinkConnectorConfig.KEY_IGNORE_CONFIG, "true");
+        props.put(ElasticsearchSinkConnectorConfig.DROP_INVALID_MESSAGE_CONFIG, "true");
+
+        elasticsearchSinkTask.start(props, failingClient);
+        elasticsearchSinkTask.put(Collections.singletonList(new SinkRecord("topic", 0, null, null, null, "foo", 0)));
+
+        // test that flush throws an exception
+        assertThrows(ConnectException.class, () -> elasticsearchSinkTask.flush(Collections.emptyMap()));
+        assertEquals(1, apiCallCounter.get());
+
+        // test that the exception is raised also if we try to add another record
+        try {
+            elasticsearchSinkTask.put(Collections.singletonList(
+                new SinkRecord("topic", 0, null, null, null, "bar", 0))
+            );
+            fail();
+        } catch (final ConnectException e) {
+            // the atomic variable isn't changed, the first exception is raised. The connector
+            // instance should die after the first exception since kafka connect will restart it
+            assert apiCallCounter.get() == 1;
+        }
```

Comment:
```suggestion
        final ConnectException e = assertThrows(ConnectException.class, () -> elasticsearchSinkTask.put(Collections.singletonList(
                new SinkRecord("topic", 0, null, null, null, "bar", 0))
            ));
        // the atomic variable isn't changed, the first exception is raised. The connector
        // instance should die after the first exception since kafka connect will restart it
        assertEquals(1, apiCallCounter.get());
```